asserts/snapasserts/validation_sets.go

```diff
@@ -383,194 +527,416 @@ func (v *ValidationSets) Add(valset *asserts.ValidationSet) error {
-						missing[rc.Name] = make(map[snap.Revision]map[string]bool)
-					}
-					if missing[rc.Name][rev] == nil {
-						missing[rc.Name][rev] = make(map[string]bool)
-					}
-					missing[rc.Name][rev][rc.validationSetKey] = true
+				missingSets[requiredRev][rc.validationSetKey] = true
+			case constraintConformityWrongRevision:
+				if wrongrevSets[requiredRev] == nil {
+					wrongrevSets[requiredRev] = make(map[string]bool)
 				}
+				wrongrevSets[requiredRev][rc.validationSetKey] = true
+			case constraintConformityValid:
+				continue
+			default:
+				return nil, nil, nil, fmt.Errorf("internal error: unknown conformity %d", conformity)
 			}
 		}
 	}
 
-	setsToLists := func(in map[string]map[string]bool) map[string][]string {
+	setToLists := func(in map[string]bool) []string {
 		if len(in) == 0 {
 			return nil
 		}
-		out := make(map[string][]string)
-		for snap, sets := range in {
-			out[snap] = make([]string, 0, len(sets))
-			for validationSetKey := range sets {
-				out[snap] = append(out[snap], validationSetKey)
-			}
-			sort.Strings(out[snap])
+		out := make([]string, 0, len(in))
+		for key := range in {
+			out = append(out, key)
 		}
+		sort.Strings(out)
 		return out
 	}
 
-	if len(invalid) > 0 || len(missing) > 0 || len(wrongrev) > 0 {
-		verr := &ValidationSetsValidationError{
-			InvalidSnaps: setsToLists(invalid),
-			Sets:         v.sets,
-		}
-		if len(missing) > 0 {
-			verr.MissingSnaps = make(map[string]map[snap.Revision][]string)
-			for snapName, revs := range missing {
-				verr.MissingSnaps[snapName] = make(map[snap.Revision][]string)
-				for rev, keys := range revs {
-					for key := range keys {
-						verr.MissingSnaps[snapName][rev] = append(verr.MissingSnaps[snapName][rev], key)
-					}
-					sort.Strings(verr.MissingSnaps[snapName][rev])
-				}
+	invalid = setToLists(invalidSet)
+	missing = make(map[snap.Revision][]string)
+	for rev, set := range missingSets {
+		missing[rev] = setToLists(set)
+	}
+
+	wrongrev = make(map[snap.Revision][]string)
+	for rev, set := range wrongrevSets {
+		wrongrev[rev] = setToLists(set)
+	}
+
+	return invalid, missing, wrongrev, nil
+}
+
+// CheckInstalledSnaps checks installed snaps against the validation sets.
+func (v *ValidationSets) CheckInstalledSnaps(snaps []*InstalledSnap, ignoreValidation map[string]bool) error {
+	installed := newInstalledSnapSet(snaps)
+	snapInstalled := func(cstrs constraints) (snap.Revision, error) {
+		if cstrs.snapRef == nil {
+			return snap.Revision{}, errors.New("internal error: snap constraint should have a snap ref")
+		}
+
+		sn := installed.Lookup(cstrs.snapRef)
+		if sn == nil {
+			return snap.R(0), nil
+		}
+		return sn.Revision, nil
+	}
+
+	snapConstraints := make([]constraints, 0, len(v.snaps))
+	for _, sc := range v.snaps {
+		snapConstraints = append(snapConstraints, sc.constraints)
+	}
+
+	invalid, missing, wrongrev, err := checkManyConstraints(snapConstraints, snapInstalled, ignoreValidation)
+	if err != nil {
```

Comment:
can this be anything but an internal error?

Reply:
Yes, they're both internal errors.